File: routine-api/src/main/java/at/enfilo/def/routine/ObjectiveRoutine.java
package at.enfilo.def.routine;

import at.enfilo.def.communication.dto.ServiceEndpointDTO;
import at.enfilo.def.communication.exception.ClientCommunicationException;
import at.enfilo.def.communication.exception.ClientCreationException;
import at.enfilo.def.datatype.DEFString;
import at.enfilo.def.parameterserver.api.IParameterServerServiceClient;
import at.enfilo.def.parameterserver.api.client.ParameterServerServiceClientFactory;
import at.enfilo.def.parameterserver.api.protocol.PrimitiveProtocolParser;
import at.enfilo.def.routine.api.Command;
import at.enfilo.def.routine.api.Order;
import at.enfilo.def.transfer.dto.ParameterProtocol;
import at.enfilo.def.transfer.dto.ProgramDTO;
import at.enfilo.def.transfer.dto.ResourceDTO;
import org.apache.thrift.TBase;
import org.apache.thrift.TException;
import org.apache.thrift.TSerializer;

import java.io.IOException;

/**
 * ObjectiveRoutine base.
 *
 * Provides an API for library development.
 *
 * To implement: run()
 * Accessing input paramenters: getInParameter()
 * Store result: setResult()
 *
 * @param <R> Result Type
 */
public abstract class ObjectiveRoutine<R extends TBase> extends AbstractRoutine {
	private final TSerializer serializer = new TSerializer();
	private String pId;
	private IParameterServerServiceClient parameterServerServiceClient;

	protected ObjectiveRoutine() {
	}

	/**
	 * Storing result of library routine.
	 *
	 * @param result - object to store
	 * @throws IOException
	 */
	private void setResult(R result) throws TException {
		log(LogLevel.DEBUG, String.format("Storing Result %s", result.toString()));
		byte[] data = serializer.serialize(result);
		out.store(data.length);
		out.store(data);
	}

	/**
	 * Fetch input parameter by given index.
	 *
	 * @param cls - class of in parameter
	 * @param name - name of parameter
	 * @param <T> - in parameter type
	 * @return - in parameter data
	 * @throws AccessParameterException
	 */
	public <T extends TBase> T getParameter(String name, Class<T> cls) throws AccessParameterException {
		// Request Parameter from WorkerService
		log(LogLevel.DEBUG, String.format("Requesting Parameter %s", name));
		Order getParam = new Order(Command.GET_PARAMETER, name);
		try {
			ctrl.store(getParam);
			T param = in.read(cls.newInstance());
			log(LogLevel.DEBUG, String.format("Received: %s", param));
			return param;

		} catch (IllegalAccessException | InstantiationException | TException e) {
			throw new AccessParameterException(e);
		}
	}

	private IParameterServerServiceClient getParameterServerClient() throws AccessParameterException {
		try {
			ServiceEndpointDTO endpoint = getParameter("parameterServerEndpoint", ServiceEndpointDTO.class);
			return new ParameterServerServiceClientFactory().createClient(endpoint);
		} catch (ClientCreationException e) {
			throw new AccessParameterException(e);
		}
	}

	@Override
	public void runRoutine() {
		// Start routine and store result
		try {
			pId = getParameter("program", DEFString.class).getValue();
			parameterServerServiceClient = getParameterServerClient();
			R result = routine();
			setResult(result);

		} catch (TException | AccessParameterException e) {
			throw new RoutineException(e);
		}
	}

	/**
	 * ObjectiveRoutine.
	 * Real implementation.
	 *
	 * @return - return parameter
	 * @throws RoutineException
	 */
	protected abstract R routine() throws RoutineException;

	/**
	 * Entry point for ObjectiveRoutine.
	 * @param args
	 */
	public static void main(String[] args) {
		if (args.length != 4) {
			String errMsg = "4 Arguments needed: <routineName> <inPipe> <outPipe> <ctrlPipe>";
			throw new RoutineException(errMsg);

		} else {

			try {
				// Instantiate routine and run
				AbstractRoutine routine = createRoutine(args[0], args[1], args[2], args[3]);
				routine.run();

			} catch (Exception e) {
				System.err.println(String.format("%s - Error while running ObjectiveRoutine: %s", LogLevel.ERROR, e.getMessage()));
				throw new RoutineException(e);
			}
		}
	}

	public float[] getFloatParameter(String name) throws AccessParameterException {
		try {
			ResourceDTO resourceDTO = parameterServerServiceClient.getParameter(pId, name, ParameterProtocol.PRIMITIVE).get();
			return (float[]) new PrimitiveProtocolParser().decode(resourceDTO);
		} catch (Exception e) {
			throw new AccessParameterException(e);
		}
	}

	public String setParameter(String name, float[] data) throws AccessParameterException {
		try {
			ResourceDTO resourceDTO = new PrimitiveProtocolParser().encode(data, "float");
			return parameterServerServiceClient.setParameter(pId, name, resourceDTO, ParameterProtocol.PRIMITIVE).get();
		} catch (Exception e) {
			throw new AccessParameterException(e);
		}
	}


	public String deleteParameter(String name) throws ClientCommunicationException {
		try {
			return parameterServerServiceClient.deleteParameter(pId, name).get();
		} catch (Exception e) {
			throw new ClientCommunicationException(e);
		}
	}

	public String deleteAllParameters() throws ClientCommunicationException {
		try {
			return parameterServerServiceClient.deleteAllParameters(pId).get();
		} catch (Exception e) {
			throw new ClientCommunicationException(e);
		}
	}

	public String addToParameter(String name, float[] data) throws AccessParameterException {
		try {
			ResourceDTO resourceDTO = new PrimitiveProtocolParser().encode(data, "float");
			return parameterServerServiceClient.addToParameter(pId, name, resourceDTO, ParameterProtocol.PRIMITIVE).get();
		} catch (Exception e) {
			throw new AccessParameterException(e);
		}
	}
}

File: apis/csharp/routine-api/ObjectiveRoutine.cs
using System;
using System.IO;
using routine_api.exception;
using common.thrift;
using Thrift;
using Thrift.Protocol;

namespace routine_api
{
    /// <inheritdoc />
    /// <summary>
    /// Defines the base functionality of an Objective Routine.
    /// </summary>
    /// <typeparam name="TR">Defines the type of the result value, which needs to be a Thrift generated type</typeparam>
    public abstract class ObjectiveRoutine<TR> : AbstractRoutine where TR : TBase
    {

        /// <summary>
        /// Stores the result of the routine implementation to the output pipe.
        /// </summary>
        /// <param name="result">The result to store</param>
        private void SetResult(TR result)
        {
            Log(LogLevel.Debug, $"Storing Result {result.ToString()}");
            byte[] data = ThriftSerializer.Serialize(result);
            OutPipe.Store(data.Length);
            OutPipe.Store(data);
        }

        /// <summary>
        /// Fetch an input parameter from the input pipe by its name. A command will be sent to the control pipe to
        /// feed the parameter to the input pipe, which will be read and mapped to the specific type T.
        /// </summary>
        /// <param name="name">The name of the input parameter to search for</param>
        /// <typeparam name="T">The type of the parameter to map to, which needs to be a Thrift generated type</typeparam>
        /// <returns>Returns an instance of type T created from the parameter</returns>
        /// <exception cref="AccessParameterException">If there is an error reading the parameter</exception>
        protected T GetParameter<T>(string name) where T : TBase
        {
            Log(LogLevel.Debug, $"Requesting Parameter {name}");

            Order getParam = new Order
            {
                Command = Command.GET_PARAMATER,
                Value = name
            };
            try
            {
                CtrlPipe.Store(getParam);
                T param = InPipe.Read(Activator.CreateInstance<T>());
                Log(LogLevel.Debug, $"Received: {param}");
                return param;
            }
            catch (Exception e)
            {
                throw new AccessParameterException(e);
            }
        }

        /// <inheritdoc />
        /// <summary>
        /// Runs the routine implementation and stores the result.
        /// </summary>
        /// <exception cref="T:routine_api.exception.RoutineException">If there is an error running the routine or storing the result</exception>
        protected override void RunRoutine()
        {
            try
            {
                TR result = Routine();
                SetResult(result);
            }
            catch (Exception e) when (e is TException || e is IOException)
            {
                throw new RoutineException(e);
            }
        }

        /// <summary>
        /// Implementation of the routine. This method will be implemented by the Algorithm developer and contains the routine logic.
        /// </summary>
        /// <returns>Returns an instance of type TR, which is the result of the routine</returns>
        protected abstract TR Routine();
    }
}